service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```
@@ -0,0 +1,104 @@
+
+    public CertificateServiceImpl(CertificateRepository certificateRepository, TagRepository tagRepository) {
+        this.certificateRepository = certificateRepository;
+        this.tagRepository = tagRepository;
+    }
+
+    @Override
+    public List<CertificateDto> findCertificateByService(String tagName, String certificateName
+            , String certificateDescription, SortType sortByName, SortType sortByCreateDate) {
+
+        if (tagName.isEmpty() && certificateName.isEmpty() && certificateDescription.isEmpty()
+                && sortByName == null && sortByCreateDate == null) {
+            throw new EntityException("There are no parameters to find a certificate");
+        }
+
+        return certificateRepository.find(tagName, certificateName
+                        , certificateDescription, sortByName, sortByCreateDate).stream()
+                .map(certificate -> {
+                    long certificateId = certificate.getId();
+                    List<Tag> tagList = tagRepository.findByCertificateId(certificateId);
+                    return CertificateMapper.toCertificateDto(certificate, tagList);
+                }).collect(Collectors.toList());
+
+    }
+
+    @Override
+    public CertificateDto findCertificateByIdService(long id) {
+        if (id < 1) {
+            throw new EntityException("There is no entity by this id: " + id);
+        }
+        return CertificateMapper.toCertificateDto(certificateRepository.findById(id)
+                .orElseThrow(() -> new EntityException("There is no entity by this id: " + id)), List.of());
+    }
+
+    @Override
+    public boolean attachTagToCertificateService(long certificateId, long tagId) {
+        if (certificateId < 1 || tagId < 1) {
+            return false;
+        }
+        return certificateRepository.attachTag(certificateId, tagId);
+    }
+
+    @Override
+    public boolean detachTagFromCertificateService(long certificateId, long tagId) {
```

What will be happened with tag which is not attached to any certificate? Will it be deleted from DB?

---

There is an AND in the sql statement for this method, if there is no match between the id's, the result would be nothing, therefore as I understand, I should throw an exception?

---

Я имею ввиду, что будет порисходить в случае, если ты отвяжешь тэг от всех сертификатов? Он будет лежать в базе как ненужная запись? Не думал ли подчищать из базы такие записи как-нибудь?

---

fixed